### ISSUE TITLE
MTC4BT: Locos do not resume their speed after all locos are connected

### ISSUE THREAD

**Issue by Mattzobricks** [Mattzobricks/MattzoControllers#16]
Locos do not resume their speed when initially connected or have reconnected after a partial or full disconnect. An additional Rocrail command is required to get the loco started - or stopped in certain situations.

To overcome this problem, you could set the presently desired speed for the loco right after connecting to all BLE hubs of that loco. For this reason, it is required that you actually store the last received loco command within the MTC4BT for each loco, or, even better, keep the desired speed of a loco in the loco object, and keep it up-to-date whenever you receive a Rocrail command - no matter if the locos of the train are fully connected or not.

Of course, all other aspects of moving or stopping trains shall be obeyed, especially the power-off feature. If Rocrail general power is off, the train must not move.

Related issues:
#19 (same for lights).
#20 (behaviour after partial disconnect).

**Comment by Mattzobricks** [comment]
I would like to extend the scope of this issue. When a train has been connected, and then partially or completely disconnects, it does not receive any train commands. Only after the reconnection of all locos, it accepts train commands again. This is critical in situations were a loco disconnects just before hitting a sensor that would make it stop. The train will actually stop because of the partial disconnect, but then start driving again after reconnection, because it missed the stop command.

**Comment by Mattzobricks** [comment]
Description of the issue modified (extended scope by the reconnection case).

**Comment by raymondbrink** [comment]
Please create a new issue as the old one is implemented and merged already.

Op za 24 dec. 2022 13:27 schreef Mattzobricks ***@***.***>:

**Comment by Mattzobricks** [comment]
I just tested it. The feature does not appear to work.

Test set-up:
- booted MTC4BT controller. Waited 20s.
- started train in rocrail (0 -> 100% power)
- powered up train
- train does not start to drive until you issue a new drive command.

**Comment by raymondbrink** [comment]
My Rocrail is dead at the moment, so I can't test it. I've pushed a branch (https://github.com/Mattzobricks/MattzoControllers/tree/16-mtc4bt-resume-speed-after-all-locos-reconnect) for testing. Could anyone give this a try...?

**Comment by Mattzobricks** [comment]
Thank you... and sure, I can help testing. I am presently on vacation and have no PU hubs or SBrick in the vicinity, but I will do so as soon as possible once I am back.

**Comment by Mattzobricks** [comment]
Tested with a Mattzes ICE train with 2 locos (SBricks) and bidirectional auto lights (two standard PF light sets in each loco).

Test 1
Setup:
- Switch both locos off.
- Boot controller
- Set a speed for the train in Rocrail.
- Switch locos on.

Results:
- Locos connect successfully after some seconds (lights blink 3 times).
- After the second loco has connected, both motors start turning correctly.
- Auto light command (e.g. lights forward) is not executed correctly. All lights remain dark.

Test 2
Setup:
- Switch one loco on, the other one off.
- Boot controller
- Wait for the first loco to connect
- Set a speed for the train in Rocrail.
- Switch the second loco on.

Results:
- Second loco connects successfully after some seconds (lights blink 3 times).
- After the second train has connected, both motors start turning correctly.
- Auto light command (e.g. lights forward) is not executed correctly. All lights remain dark.

Test 3
Setup:
- Switch both locoso on.
- Boot controller
- Set a speed for the train in Rocrail, so that both motors turn.
- Switch second loco off.
- Wait 10 seconds.
- Switch second loco back on.

Results:
- After the second loco has disconnected, the first loco stops immediately.
- After the second loco has reconnected, both motors restart turning correctly.
- Auto light command (e.g. lights forward) is not executed correctly. All lights remain dark.

Test 4:
Setup:
- Like test 3, both with switching both locos off and on again.

Results:
- Works as desired, except for auto lights (as in test 3).


Total result: nice job, but we need to deal with the auto light problem.

**Comment by Mattzobricks** [comment]
I would recommend to merge this branch into develop.

**Comment by raymondbrink** [comment]
Pull request submitted. This issue was never about lights, so didn't give it any attention. Will do when I start on #19.

**Comment by raymondbrink** [comment]
@Mattzobricks As this is now merged into develop, shouldn't this issue be closed?

**Comment by Mattzobricks** [comment]
Yes, thank you.
